@here/harp-mapview/lib/BoundsGenerator.ts

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import {
+    EarthConstants,
+    GeoCoordinates,
+    mercatorTilingScheme,
+    ProjectionType,
+    TileKey
+} from "@here/harp-geoutils";
+import { Frustum, Line3, Matrix4, Plane, Vector2, Vector3 } from "three";
+
+import { TileGeometryCreator } from "./geometry/TileGeometryCreator";
+import { MapView } from "./MapView";
+import { MapViewUtils } from "./Utils";
+
+export class BoundsGenerator {
+    //TODO: check what we really need here
+    constructor(private m_mapView: MapView) {}
+
+    generate(): GeoCoordinates[] {
+        //TODO: support spherical projection
+
+        //!!!!!!!ALTITUDE IS NOT TAKEN INTO ACCOUNT!!!!!!!!!
+        const geoPolygon: GeoCoordinates[] = [];
+
+        //CASE A: FLAT PROJECTION
+        if (this.m_mapView.projection.type === ProjectionType.Planar) {
+            // 1.) Raycast into all for corners of the canvas
+            //     => if an intersection is found, add it to the polygon
+            this.raycastCanvasCorners(geoPolygon);
```

Comment:
Not sure what we want to return in case not all rays hit the surface but wouldn't it work if we fallback to raycasting the far-plane in that case? That would reduce all the special case handling below.

Reply:
Not sure I understand your point, in case no surface is hit, there would be no return value, as its not part of the searched polygon. 
What do you mean with raycasting the far-plane, raycast the CanvasCorners against the farplane will result in the back corners of the frustum, the intersection of the groundPlane with the farplane is done with the next check looking for the horizon,